File: frameworks/go-graphql-go/queries_test.go
package main

import (
	"testing"

	"github.com/benchmark/go-graphql-go/internal/graphql"
)

// ============================================================================
// Query: User Operations
// ============================================================================

func TestUserQueries(t *testing.T) {
	tests := []struct {
		name      string
		username  string
		fullName  string
		bio       string
		expectErr bool
	}{
		{
			name:      "user query returns valid user",
			username:  "alice",
			fullName:  "Alice",
			bio:       "Developer",
			expectErr: false,
		},
		{
			name:      "user with null bio",
			username:  "bob",
			fullName:  "Bob",
			bio:       "",
			expectErr: false,
		},
		{
			name:      "user with special characters",
			username:  "charlie",
			fullName:  "Char'lie O'Neill",
			bio:       `Quote: "Hello & goodbye"`,
			expectErr: false,
		},
		{
			name:      "user with unicode",
			username:  "diana",
			fullName:  "Diàna Müller",
			bio:       "José García",
			expectErr: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			user := factory.CreateTestUser(tt.username, tt.username+"@example.com", tt.fullName, tt.bio)

			// Act
			retrieved := factory.GetUser(user.ID)

			// Assert
			if retrieved == nil {
				t.Errorf("User not found: %s", user.ID)
				return
			}
			if retrieved.ID != user.ID {
				t.Errorf("ID mismatch: expected %s, got %s", user.ID, retrieved.ID)
			}
			if retrieved.Username != tt.username {
				t.Errorf("Username mismatch: expected %s, got %s", tt.username, retrieved.Username)
			}
		})
	}
}

// ============================================================================
// Query: Post Operations
// ============================================================================

func TestPostQueries(t *testing.T) {
	tests := []struct {
		name     string
		title    string
		content  string
		withAuthor bool
	}{
		{
			name:     "post query returns valid post",
			title:    "Test Post",
			content:  "Test Content",
			withAuthor: true,
		},
		{
			name:     "post with null content",
			title:    "No Content Post",
			content:  "",
			withAuthor: true,
		},
		{
			name:     "post with special characters",
			title:    "Post with 'quotes' & <html>",
			content:  "Content: \"test\" & special",
			withAuthor: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			var authorID string
			if tt.withAuthor {
				author := factory.CreateTestUser("author", "author@example.com", "Author", "")
				authorID = author.ID
			}

			post := factory.CreateTestPost(authorID, tt.title, tt.content)

			// Act
			retrieved := factory.GetPost(post.ID)

			// Assert
			if retrieved == nil {
				t.Errorf("Post not found: %s", post.ID)
				return
			}
			if retrieved.Title != tt.title {
				t.Errorf("Title mismatch: expected %s, got %s", tt.title, retrieved.Title)
			}
		})
	}
}

// ============================================================================
// Query: List Operations
// ============================================================================

func TestListQueries(t *testing.T) {
	tests := []struct {
		name      string
		count     int
		limit     int
		minExpect int
	}{
		{
			name:      "list users returns all",
			count:     5,
			limit:     10,
			minExpect: 5,
		},
		{
			name:      "list respects limit",
			count:     20,
			limit:     5,
			minExpect: 5,
		},
		{
			name:      "empty list returns empty",
			count:     0,
			limit:     10,
			minExpect: 0,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			for i := 0; i < tt.count; i++ {
				factory.CreateTestUser(
					"user"+string(rune(48+i%10)),
					"user"+string(rune(48+i%10))+"@example.com",
					"User "+string(rune(48+i%10)),
					"",
				)
			}

			// Act
			users := factory.GetAllUsers()

			// Assert
			if len(users) < tt.minExpect {
				t.Errorf("Expected at least %d users, got %d", tt.minExpect, len(users))
			}
		})
	}
}

// ============================================================================
// Query: Relationship Queries
// ============================================================================

func TestRelationshipQueries(t *testing.T) {
	tests := []struct {
		name         string
		postsPerUser int
	}{
		{
			name:         "user includes posts",
			postsPerUser: 3,
		},
		{
			name:         "post includes author",
			postsPerUser: 1,
		},
		{
			name:         "user with no posts",
			postsPerUser: 0,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			author := factory.CreateTestUser("author", "author@example.com", "Author", "")

			for i := 0; i < tt.postsPerUser; i++ {
				factory.CreateTestPost(author.ID, "Post "+string(rune(48+i)), "Content")
			}

			// Act
			posts := factory.GetAllPosts()

			// Assert
			userPostCount := 0
			for _, post := range posts {
				if post.AuthorID != nil && *post.AuthorID == author.ID {
					userPostCount++
				}
			}

			if userPostCount != tt.postsPerUser {
				t.Errorf("Expected %d posts for author, got %d", tt.postsPerUser, userPostCount)
			}
		})
	}
}

// ============================================================================
// Query: Trinity Identifier Pattern
// ============================================================================

func TestTrinityIdentifierPattern(t *testing.T) {
	tests := []struct {
		name string
	}{
		{
			name: "all user IDs are valid UUIDs",
		},
		{
			name: "all post IDs are valid UUIDs",
		},
		{
			name: "post author ID references user",
		},
		{
			name: "multiple users have unique IDs",
		},
		{
			name: "multiple posts have unique IDs",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			validator := NewValidationHelper(t)
			defer factory.Reset()

			if tt.name == "all user IDs are valid UUIDs" {
				user := factory.CreateTestUser("alice", "alice@example.com", "Alice", "")
				validator.AssertUUID(user.ID)
			} else if tt.name == "all post IDs are valid UUIDs" {
				author := factory.CreateTestUser("author", "author@example.com", "Author", "")
				post := factory.CreateTestPost(author.ID, "Test", "Content")
				validator.AssertUUID(post.ID)
			} else if tt.name == "post author ID references user" {
				author := factory.CreateTestUser("author", "author@example.com", "Author", "")
				post := factory.CreateTestPost(author.ID, "Test", "Content")
				if post.AuthorID == nil || *post.AuthorID != author.ID {
					t.Errorf("Post author ID does not match user ID")
				}
			} else if tt.name == "multiple users have unique IDs" {
				user1 := factory.CreateTestUser("alice", "alice@example.com", "Alice", "")
				user2 := factory.CreateTestUser("bob", "bob@example.com", "Bob", "")
				if user1.ID == user2.ID {
					t.Errorf("User IDs should be unique")
				}
			} else if tt.name == "multiple posts have unique IDs" {
				author := factory.CreateTestUser("author", "author@example.com", "Author", "")
				post1 := factory.CreateTestPost(author.ID, "Post1", "Content1")
				post2 := factory.CreateTestPost(author.ID, "Post2", "Content2")
				if post1.ID == post2.ID {
					t.Errorf("Post IDs should be unique")
				}
			}
		})
	}
}

// ============================================================================
// Query: Data Consistency
// ============================================================================

func TestDataConsistency(t *testing.T) {
	tests := []struct {
		name string
	}{
		{
			name: "same user returns consistent data",
		},
		{
			name: "same post returns consistent data",
		},
		{
			name: "multiple queries return same results",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			if tt.name == "same user returns consistent data" {
				user := factory.CreateTestUser("alice", "alice@example.com", "Alice", "Bio")
				retrieved1 := factory.GetUser(user.ID)
				retrieved2 := factory.GetUser(user.ID)

				if retrieved1 == nil || retrieved2 == nil {
					t.Errorf("User not found")
					return
				}
				if retrieved1.ID != retrieved2.ID {
					t.Errorf("User ID mismatch on multiple queries")
				}
				if retrieved1.Username != retrieved2.Username {
					t.Errorf("Username changed between queries")
				}
			} else if tt.name == "same post returns consistent data" {
				author := factory.CreateTestUser("author", "author@example.com", "Author", "")
				post := factory.CreateTestPost(author.ID, "Test", "Content")
				retrieved1 := factory.GetPost(post.ID)
				retrieved2 := factory.GetPost(post.ID)

				if retrieved1 == nil || retrieved2 == nil {
					t.Errorf("Post not found")
					return
				}
				if retrieved1.Title != retrieved2.Title {
					t.Errorf("Post title changed between queries")
				}
			}
		})
	}
}

// ============================================================================
// Query: Null Field Handling
// ============================================================================

func TestNullFieldHandling(t *testing.T) {
	tests := []struct {
		name        string
		bio         string
		wantNil     bool
	}{
		{
			name:        "null bio is handled",
			bio:         "",
			wantNil:     true,
		},
		{
			name:        "present bio is returned",
			bio:         "My bio",
			wantNil:     false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			user := factory.CreateTestUser("alice", "alice@example.com", "Alice", tt.bio)

			// Act
			retrieved := factory.GetUser(user.ID)

			// Assert
			if retrieved == nil {
				t.Errorf("User not found")
				return
			}

			if tt.wantNil {
				if retrieved.Bio != nil {
					t.Errorf("Expected nil bio, got %v", retrieved.Bio)
				}
			} else {
				if retrieved.Bio == nil {
					t.Errorf("Expected non-nil bio")
				}
			}
		})
	}
}

// ============================================================================
// Query: Special Characters
// ============================================================================

func TestSpecialCharacterQueries(t *testing.T) {
	tests := []struct {
		name    string
		content string
	}{
		{
			name:    "handles single quotes",
			content: "I'm a developer",
		},
		{
			name:    "handles double quotes",
			content: `He said "hello"`,
		},
		{
			name:    "handles HTML tags",
			content: "Check <this> out",
		},
		{
			name:    "handles ampersand",
			content: "Tom & Jerry",
		},
		{
			name:    "handles emoji",
			content: "🎉 Celebration! 🚀 Rocket",
		},
		{
			name:    "handles unicode accents",
			content: "Àlice Müller",
		},
		{
			name:    "handles diacritics",
			content: "José García",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			// Act
			post := factory.CreateTestPost("user-id", "Title", tt.content)

			// Assert
			if post == nil {
				t.Errorf("Post creation failed")
				return
			}

			retrieved := factory.GetPost(post.ID)
			if retrieved == nil {
				t.Errorf("Post not found")
				return
			}

			if retrieved.Content != nil {
				if *retrieved.Content != tt.content {
					t.Errorf("Content mismatch: expected %q, got %q", tt.content, *retrieved.Content)
				}
			}
		})
	}
}

// ============================================================================
// Query: Boundary Conditions
// ============================================================================

func TestBoundaryConditionQueries(t *testing.T) {
	tests := []struct {
		name      string
		size      int
		limit     int
	}{
		{
			name:      "limit 0 returns empty",
			size:      10,
			limit:     0,
		},
		{
			name:      "limit 1 returns one",
			size:      10,
			limit:     1,
		},
		{
			name:      "limit larger than size returns all",
			size:      5,
			limit:     100,
		},
		{
			name:      "very large limit handled",
			size:      10,
			limit:     999999,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			for i := 0; i < tt.size; i++ {
				factory.CreateTestUser(
					"user"+string(rune(48+i%10)),
					"user"+string(rune(48+i%10))+"@example.com",
					"User",
					"",
				)
			}

			// Act
			users := factory.GetAllUsers()

			// Assert
			if len(users) < tt.size {
				t.Errorf("Expected at least %d users, got %d", tt.size, len(users))
			}
		})
	}
}

// ============================================================================
// Query: Multi-Author Posts
// ============================================================================

func TestMultiAuthorQueries(t *testing.T) {
	tests := []struct {
		name        string
		authorCount int
		postsPerAuthor int
	}{
		{
			name:        "multiple authors have separate posts",
			authorCount: 3,
			postsPerAuthor: 2,
		},
		{
			name:        "author isolation verified",
			authorCount: 2,
			postsPerAuthor: 3,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			// Arrange
			factory := NewTestFactory(t)
			defer factory.Reset()

			authorPostCounts := make(map[string]int)

			for i := 0; i < tt.authorCount; i++ {
				author := factory.CreateTestUser(
					"author"+string(rune(48+i)),
					"author"+string(rune(48+i))+"@example.com",
					"Author "+string(rune(48+i)),
					"",
				)

				for j := 0; j < tt.postsPerAuthor; j++ {
					factory.CreateTestPost(author.ID, "Post "+string(rune(48+j)), "Content")
					authorPostCounts[author.ID]++
				}
			}

			// Act & Assert
			for authorID, expectedCount := range authorPostCounts {
				if expectedCount != tt.postsPerAuthor {
					t.Errorf("Author %s: expected %d posts, got %d", authorID, tt.postsPerAuthor, expectedCount)
				}
			}
		})
	}
}

File: frameworks/juniper/tests/queries_test.rs
mod test_helpers;
use test_helpers::*;

// ============================================================================
// Query: Ping
// ============================================================================

#[test]
fn test_ping_query() {
    assert!(true);
}

// ============================================================================
// Query: User List
// ============================================================================

#[test]
fn test_user_list_query() {
    let factory = TestFactory::new();
    factory.create_test_user("alice", "alice@example.com", "Alice", "");
    factory.create_test_user("bob", "bob@example.com", "Bob", "");
    factory.create_test_user("charlie", "charlie@example.com", "Charlie", "");

    let users = factory.get_all_users();
    assert_eq!(users.len(), 3);
}

#[test]
fn test_user_list_with_limit() {
    let factory = TestFactory::new();
    for i in 0..20 {
        factory.create_test_user(
            &format!("user{}", i),
            &format!("user{}@example.com", i),
            "User",
            "",
        );
    }

    let users = factory.get_all_users();
    assert!(users.len() >= 20);
}

#[test]
fn test_user_list_empty() {
    let factory = TestFactory::new();
    let users = factory.get_all_users();
    assert_eq!(users.len(), 0);
}

#[test]
fn test_user_list_with_pagination() {
    let factory = TestFactory::new();
    for i in 0..30 {
        factory.create_test_user(
            &format!("user{}", i % 10),
            &format!("user{}@example.com", i % 10),
            "User",
            "",
        );
    }

    let users = factory.get_all_users();
    assert!(users.len() >= 10);
}

// ============================================================================
// Query: User Detail
// ============================================================================

#[test]
fn test_user_detail_query() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "Developer");
    let user_id = user.id.clone();

    let retrieved = factory.get_user(&user_id);
    assert!(retrieved.is_some());
    assert_eq!(retrieved.unwrap().username, "alice");
}

#[test]
fn test_user_detail_with_bio() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "My bio");
    let user_id = user.id.clone();

    let retrieved = factory.get_user(&user_id);
    assert!(retrieved.is_some());
    assert_eq!(retrieved.unwrap().bio, Some("My bio".to_string()));
}

#[test]
fn test_user_detail_without_bio() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("bob", "bob@example.com", "Bob", "");
    let user_id = user.id.clone();

    let retrieved = factory.get_user(&user_id);
    assert!(retrieved.is_some());
    assert_eq!(retrieved.unwrap().bio, None);
}

#[test]
fn test_user_detail_not_found() {
    let factory = TestFactory::new();
    let user = factory.get_user("nonexistent-id");
    assert!(user.is_none());
}

// ============================================================================
// Query: Post List
// ============================================================================

#[test]
fn test_post_list_query() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");

    factory.create_test_post(&author.id, "Post 1", "Content 1");
    factory.create_test_post(&author.id, "Post 2", "Content 2");
    factory.create_test_post(&author.id, "Post 3", "Content 3");

    let posts = factory.get_all_posts();
    assert_eq!(posts.len(), 3);
}

#[test]
fn test_post_list_with_limit() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");

    for i in 0..20 {
        factory.create_test_post(&author.id, &format!("Post {}", i), "Content");
    }

    let posts = factory.get_all_posts();
    assert!(posts.len() >= 20);
}

#[test]
fn test_post_list_empty() {
    let factory = TestFactory::new();
    let posts = factory.get_all_posts();
    assert_eq!(posts.len(), 0);
}

// ============================================================================
// Query: Post Detail
// ============================================================================

#[test]
fn test_post_detail_query() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post = factory.create_test_post(&author.id, "Test Post", "Test Content");
    let post_id = post.id.clone();

    let retrieved = factory.get_post(&post_id);
    assert!(retrieved.is_some());
    let retrieved = retrieved.unwrap();
    assert_eq!(retrieved.title, "Test Post");
    assert_eq!(retrieved.content, "Test Content");
}

#[test]
fn test_post_detail_not_found() {
    let factory = TestFactory::new();
    let post = factory.get_post("nonexistent-id");
    assert!(post.is_none());
}

// ============================================================================
// Query: Trinity Identifier Pattern
// ============================================================================

#[test]
fn test_user_has_uuid_id() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "");
    assert!(ValidationHelper::assert_uuid(&user.id).is_ok());
}

#[test]
fn test_user_has_primary_key() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "");
    assert!(user.pk_user > 0);
}

#[test]
fn test_post_has_uuid_id() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post = factory.create_test_post(&author.id, "Post", "Content");
    assert!(ValidationHelper::assert_uuid(&post.id).is_ok());
}

#[test]
fn test_post_has_primary_key() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post = factory.create_test_post(&author.id, "Post", "Content");
    assert!(post.pk_post > 0);
}

#[test]
fn test_trinity_identifiers_separate() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "");
    assert_ne!(user.id, user.pk_user.to_string());
}

// ============================================================================
// Query: User-Post Relationships
// ============================================================================

#[test]
fn test_user_has_multiple_posts() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");

    factory.create_test_post(&author.id, "Post 1", "Content");
    factory.create_test_post(&author.id, "Post 2", "Content");
    factory.create_test_post(&author.id, "Post 3", "Content");

    let author_posts: Vec<_> = factory
        .get_all_posts()
        .into_iter()
        .filter(|p| p.fk_author == author.pk_user)
        .collect();

    assert_eq!(author_posts.len(), 3);
}

#[test]
fn test_post_references_author() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post = factory.create_test_post(&author.id, "Post", "Content");

    assert_eq!(post.fk_author, author.pk_user);
}

// ============================================================================
// Query: Data Consistency
// ============================================================================

#[test]
fn test_list_and_detail_data_match() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "Bio");

    let list_user = factory.get_user(&user.id);
    let detail_user = factory.get_user(&user.id);

    assert_eq!(
        list_user.as_ref().unwrap().username,
        detail_user.as_ref().unwrap().username
    );
}

#[test]
fn test_repeated_queries_return_same_data() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "");

    let retrieved1 = factory.get_user(&user.id);
    let retrieved2 = factory.get_user(&user.id);

    assert_eq!(retrieved1.as_ref().unwrap().id, retrieved2.as_ref().unwrap().id);
}

// ============================================================================
// Query: Null Field Handling
// ============================================================================

#[test]
fn test_user_with_null_bio_in_list() {
    let factory = TestFactory::new();
    factory.create_test_user("alice", "alice@example.com", "Alice", "");
    factory.create_test_user("bob", "bob@example.com", "Bob", "Bio");

    let users = factory.get_all_users();
    assert!(users.iter().any(|u| u.bio.is_none()));
    assert!(users.iter().any(|u| u.bio.is_some()));
}

#[test]
fn test_post_with_content() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post = factory.create_test_post(&author.id, "Post", "Content");

    assert!(!post.content.is_empty());
}

// ============================================================================
// Query: Special Characters
// ============================================================================

#[test]
fn test_user_with_quotes_in_name() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Char'lie", "");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_user_with_double_quotes_in_bio() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "He said \"hello\"");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_post_with_html_tags() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post = factory.create_test_post(&author.id, "Post with <tags>", "Content");
    assert!(factory.get_post(&post.id).is_some());
}

#[test]
fn test_post_with_ampersand() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post = factory.create_test_post(&author.id, "Post", "Tom & Jerry");
    assert!(factory.get_post(&post.id).is_some());
}

#[test]
fn test_post_with_emoji() {
    let factory = TestFactory::new();
    let author = factory.create_test_user("author", "author@example.com", "Author", "");
    let post = factory.create_test_post(&author.id, "Post", "🎉 Celebration! 🚀 Rocket");
    assert!(factory.get_post(&post.id).is_some());
}

#[test]
fn test_user_with_accents() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Àlice Müller", "");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_user_with_diacritics() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "José García", "");
    assert!(factory.get_user(&user.id).is_some());
}

// ============================================================================
// Query: Boundary Conditions
// ============================================================================

#[test]
fn test_query_with_limit_0() {
    let factory = TestFactory::new();
    for i in 0..30 {
        factory.create_test_user(
            &format!("user{}", i % 10),
            &format!("user{}@example.com", i % 10),
            "User",
            "",
        );
    }

    let users = factory.get_all_users();
    assert!(users.len() > 0);
}

#[test]
fn test_query_with_limit_1() {
    let factory = TestFactory::new();
    factory.create_test_user("alice", "alice@example.com", "Alice", "");
    factory.create_test_user("bob", "bob@example.com", "Bob", "");

    let users = factory.get_all_users();
    assert!(users.len() >= 1);
}

#[test]
fn test_query_with_limit_larger_than_total() {
    let factory = TestFactory::new();
    for i in 0..5 {
        factory.create_test_user(
            &format!("user{}", i),
            &format!("user{}@example.com", i),
            "User",
            "",
        );
    }

    let users = factory.get_all_users();
    assert_eq!(users.len(), 5);
}

#[test]
fn test_query_with_very_large_limit() {
    let factory = TestFactory::new();
    for i in 0..10 {
        factory.create_test_user(
            &format!("user{}", i),
            &format!("user{}@example.com", i),
            "User",
            "",
        );
    }

    let users = factory.get_all_users();
    assert!(users.len() >= 10);
}

// ============================================================================
// Query: Multiple Entities
// ============================================================================

#[test]
fn test_multiple_users_isolated() {
    let factory = TestFactory::new();
    let user1 = factory.create_test_user("alice", "alice@example.com", "Alice", "Bio1");
    let user2 = factory.create_test_user("bob", "bob@example.com", "Bob", "Bio2");

    let retrieved1 = factory.get_user(&user1.id);
    let retrieved2 = factory.get_user(&user2.id);

    assert_ne!(retrieved1.as_ref().unwrap().id, retrieved2.as_ref().unwrap().id);
}

#[test]
fn test_multiple_authors_posts_isolated() {
    let factory = TestFactory::new();
    let author1 = factory.create_test_user("author1", "author1@example.com", "Author1", "");
    let author2 = factory.create_test_user("author2", "author2@example.com", "Author2", "");

    factory.create_test_post(&author1.id, "Post1", "Content");
    factory.create_test_post(&author2.id, "Post2", "Content");

    let author1_posts: Vec<_> = factory
        .get_all_posts()
        .into_iter()
        .filter(|p| p.fk_author == author1.pk_user)
        .collect();
    let author2_posts: Vec<_> = factory
        .get_all_posts()
        .into_iter()
        .filter(|p| p.fk_author == author2.pk_user)
        .collect();

    assert_eq!(author1_posts.len(), 1);
    assert_eq!(author2_posts.len(), 1);
}
